### PR TITLE
useAppState hook

### PR DESCRIPTION
**Pull request by dulmandakh**
## Summary

This PR introduces useAppState hook.

## Changelog

[GENERAL] [Added] - new useAppState hook

## Test Plan

It works 👍

### CODE REVIEW

**Comment by brunolemos** [comment]
There’s this repo with some more react native hooks: https://github.com/react-native-community/react-native-hooks

Are we gonna have most of them inside the official react-native package? 

Given the Lean Core initiative I was not sure, but it would make sense, because it could be used internally by react-native once it starts to use less classes.

**Comment by dulmandakh** [comment]
@brunolemos yep, that's the reason.

**Comment by TheSavior** [comment]
This might be a good opportunity, can you try adding a jest test for this hook?

**Comment by benoitdion** [comment]
related: https://github.com/facebook/react-native/issues/25127

File: Libraries/AppState/AppState.js [hunk 1]
```diff
@@ -18,6 +18,8 @@ const logError = require('../Utilities/logError');
 const invariant = require('invariant');
 
 /**
+ * NOTE: `useAppState` is the preffered API for React components.
```
Comment (cpojer) [inline Libraries/AppState/AppState.js]: *preferred.

File: Libraries/AppState/useAppState.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    AppState.addEventListener('change', handleChange)
+    setState(AppState.currentState)
+    return () => {
+      AppState.removeEventListener('change', handleChange)
+    }
+  }, [])
+  return state
+}
```
Comment (cpojer) [inline Libraries/AppState/useAppState.js]: Needs a newline

**Comment by benoitdion** [comment]
To add a bit more color to the issue I linked to above, it means that a `useAppState` hook isn't going to work properly with `useEffect`. 

You should be able to reproduce with something like the following:

```javascript
const AppStateView = () => {
  const appState = useAppState();

  useEffect(() => {
    console.log(">>> [useEffect] new app state:", appState);
  }, [appState]);

  useLayoutEffect(() => {
    console.log(">>> [useLayoutEffect] new app state:", appState);
  }, [appState]);

  return <Text>{appState}</Text>;
};
```

**Comment by radeno** [comment]
@cpojer @dulmandakh 
There is community module which has similar behavior https://github.com/react-native-community/react-native-hooks#useappstate
I think this "utility" functions should be in community module if they are not used internally in RN core.
Even RN core utility `useWindowDimensions` is reimplemented it as `useDimensions`.
I think, core should be lean and prefer that module instead implementing same features into core.

**Comment by dulmandakh** [comment]
IMHO, RN should include hooks for core modules

**Comment by radeno** [comment]
@dulmandakh i agree with you about using hooks. But it depends how to propagate it.
There are two ways:
1. Add to documentation that community module exists. It handles more that appState and screenDimensions. It is good.
2. Add hooks into every core module. It should be done for all core modules and published in one version because consistency. Later, it will be easier to extract some core modules as independed react-native-community module with same functionality.

First option is low-hanging fruit. Second option should be future-proof, but takes a lot time.

**Comment by cpojer** [comment]
We should provide hook based APIs for React Native and possibly integrate the hooks from the community project. That will open a path to possibly deprecate older APIs in the future.

File: Libraries/AppState/useAppState.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+export default function useAppState(): string {
+  const [state, setState] = useState(AppState.currentState);
+  useEffect(() => {
+    function handleChange(state) {
+      setState(state)
+    }
+
+    AppState.addEventListener('change', handleChange)
+    setState(AppState.currentState)
```
Comment (radeno) [inline Libraries/AppState/useAppState.js]: Is it necessary to set state again after mounting?

**Comment by cpojer** [comment]
@dulmandakh what's the status of this PR?

**Pull request closed without merging**